todo_app/app.py

```diff
@@ -1,59 +1,138 @@
-from flask import Flask, render_template, request, redirect
+from flask import Flask, render_template, request, redirect, flash
 
 from todo_app.flask_config import Config
 
 from todo_app.data.todo_board import ToDoBoard
 
 import os
 
+from flask_login import LoginManager, login_required, UserMixin, login_user, current_user, AnonymousUserMixin
+from oauthlib.oauth2 import WebApplicationClient
+from functools import wraps
+import requests
+
 def create_app():
     app = Flask(__name__)
     app.config.from_object('todo_app.flask_config.Config')
 
     todo_board_id = os.getenv('TODO_BOARD_ID')
+    github_client_id = os.getenv('CLIENT_ID')
+    github_client_secret = os.getenv('CLIENT_SECRET')
+
+    app.config['LOGIN_DISABLED'] = os.getenv('LOGIN_DISABLED') == 'True'
+
+    login_manager = LoginManager()
+    login_manager.anonymous_user = MyCustomAnonymousUser
+
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        client = WebApplicationClient(github_client_id)
+        return redirect(client.prepare_request_uri('https://github.com/login/oauth/authorize'))
+        
+    @login_manager.user_loader
+    def load_user(user_id):
+        return User(user_id)
 
+    login_manager.init_app(app) 
+
+    @app.route('/login/callback')
+    def login():
+        #TODO - state check
```

Comment:
If you're going to give this a go, the key considerations are:
* The state should be generated fresh for every user who's authenticating before we initially redirect them to GitHub
* Which means we need a way to recognise them again later - so you'll probably want to store it on their session
* So then here we can verify that the state passed on from GitHub matches the state that we logged for the user before we redirected them